atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,641 @@
+		data["pre_build_id"] = b.preId
+	}
+
+	if b.id != 0 {
+		data["build_id"] = b.id
+	}
+
+	if b.resourceId != 0 {
+		data["resource"] = b.resourceName
+	}
+
+	return data
+}
+
+func (b *inMemoryCheckBuild) TracingAttrs() tracing.Attrs {
+	attrs := tracing.Attrs{
+		"build":    b.Name(),
+		"team":     b.TeamName(),
+		"pipeline": b.PipelineName(),
+	}
+
+	if b.preId != 0 {
+		attrs["pre_build_id"] = fmt.Sprintf("%d", b.preId)
+	}
+
+	if b.id != 0 {
+		attrs["build_id"] = fmt.Sprintf("%d", b.id)
+	}
+
+	if b.resourceId != 0 {
+		attrs["resource"] = b.resourceName
+	}
+
+	return attrs
+}
+
+// Reload just does nothing because an in-memory build lives shortly.
+func (b *inMemoryCheckBuild) Reload() (bool, error) {
+	return true, nil
+}
+
+func (b *inMemoryCheckBuild) PrivatePlan() atc.Plan {
+	return b.plan
+}
+
+// OnCheckBuildStart is a hook point called once a check build starts. For
+// in-memory check build, this is a chance to initialize database connection.
+func (b *inMemoryCheckBuild) OnCheckBuildStart() error {
+	if !b.running {
+		return errors.New("not running build cannot start check")
+	}
+
+	if b.runningInContainer {
+		return nil
+	}
+
+	b.runningInContainer = true
+
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	if !b.dbInited {
+		err := b.initDbStuff(tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+
+	return b.conn.Bus().Notify(buildEventsChannel(b.id))
+}
+
+// AcquireTrackingLock tries to acquire a lock on checkable's ID in order to
+// avoid duplicate checks on the same checkable among ATCs and Lidar scan
+// intervals.
+func (b *inMemoryCheckBuild) AcquireTrackingLock(logger lager.Logger, interval time.Duration) (lock.Lock, bool, error) {
+	if !b.running {
+		return nil, false, errors.New("not running build cannot acquire tracking lock")
+	}
+
+	var lockId lock.LockID
+	if b.ResourceID() != 0 {
+		lockId = lock.NewInMemoryCheckBuildTrackingLockID("resource", b.ResourceID())
+	} else {
+		return nil, false, errors.New("")
+	}
+
+	lock, acquired, err := b.lockFactory.Acquire(
+		logger.Session("lock", lager.Data{}),
+		lockId,
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !acquired {
+		return nil, false, nil
+	}
+
+	return lock, true, nil
+}
+
+func (b *inMemoryCheckBuild) Finish(status BuildStatus) error {
+	if !b.running {
+		return errors.New("not running build cannot finish")
+	}
+
+	if !b.runningInContainer {
+		return nil
+	}
+
+	b.status = status
+	b.endTime = time.Now()
+
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	err = b.saveEvent(tx, event.Status{
+		Status: atc.BuildStatus(status),
+		Time:   time.Now().Unix(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Release the containers using in this build, so that they can be GC-ed.
+	_, err = psql.Update("containers").
+		Set("in_memory_build_id", nil).
+		Where(sq.Eq{"in_memory_build_id": b.preId}).
```

Comment:
In memory build only existing on the ATC where it is created. When an ATC crashes, other ATC may create new in-memory builds for the resource. So the problem you are worry doesn't exist.

I intentionally use preId because when we build a `ResourceCacheOwner`, we don't know if it will really be used. We only get a build id when in-memory inits db, and init-db only happens when a check need to really run.

I have added some GC logic today to cleanup dirty resource cache uses in case of ATC crashes.